### PR TITLE
Add minetest.bulk_set_node call + optimize Environment::set_node call

### PR DESCRIPTION
**Pull request by nerzhul**
Add bulk_set_node call permitting to modders to have a very fast interface to set nodes, removing core side node multi repetition of the node deserialization
This type of call is useful for world edit volumes to have a blazing fast node modification.

Also optimize a little bit ServerEnvironment::set_node call contentfeature lookups

### CODE REVIEW

File: doc/lua_api.txt [hunk 1]
```diff
@@ -2749,6 +2749,10 @@ and `minetest.auth_reload` call the authentication handler.
     * `node`: table `{name=string, param1=number, param2=number}`
     * If param1 or param2 is omitted, it's set to `0`.
     * e.g. `minetest.set_node({x=0, y=10, z=0}, {name="default:wood"})`
+* `minetest.bulk_set_node([pos1, pos2, posX, ...], node)`
```
Comment (sfan5) [inline doc/lua_api.txt]: <pre>
* `minetest.bulk_set_node({pos1, pos2, pos3, ...}, node)`
</pre>

**Comment by nerzhul** [comment]
I did the bench in debug mode, 45s with set_node, 40 with bulk_set_node
In relwithdebinfo: set_node loop[16620.64ms], bulk_set_node[13972.92ms]

**Comment by paramat** [comment]
Interesting feature. What advantage does it have over voxelmanip? I guess placement callbacks? Also it might be faster than LVM for smaller numbers of nodes.

**Comment by nerzhul** [comment]
@paramat yes it keep the current usage and permit massive performance improvement (the different increase with more nodes).
What do you mean by LVM ?

**Comment by paramat** [comment]
Lua Voxel Manipulator.
Nice feature.

**Comment by nerzhul** [comment]
@paramat don't hesitate to review/approve

**Comment by paramat** [comment]
This seems useful for smaller numbers of nodes, not large numbers, as a lua voxelmanip is faster for large numbers and 'set node' is unusable for large numbers even with this PR.
There was some talk of large numbers of nodes on IRC but that seems the wrong application.
This seems somewhat useful and is simple. Might review later.

**Comment by raymoo** [comment]
This is better than VM for setting sparse sets of nodes, though I don't know what use case would require sparse sets. Maybe for building hollow structures? Sort of a contrived example, but setting a NxNxN hollow cube with a 1-node thick shell should only require complexity N^2 time, while using a VM would take N^3 time due to the copying.

A `bulk_get_node` may be a good idea too. You could imagine a similar use case in detecting sparse multiblock structures.

**Comment by rubenwardy** [comment]
would be good to have benchmarks with LVM too

**Comment by paramat** [comment]
At a certain low number of nodes (10-100?), LVM becomes faster than 'set node', i see this PR as useful below that number, not as competition for LVM.

**Comment by raymoo** [comment]
@paramat What about for sparse use cases?

**Comment by paramat** [comment]
Widely separated positions? Yes that's a good use.

**Comment by nerzhul** [comment]
Then approval or not approval, that is the question :p

**Comment by paramat** [comment]
I'll try to review the bulk set node part of this within a few days.

**Comment by rubenwardy** [comment]
[graph](https://i.rubenwardy.com/IsgC63puZw.png)

<details>
To run, do:

* `/teleport 0,0,0`
* `/emergeblocks here 100`
* `/a`


```lua
assert(type(jit) ~= 'table', "Plain Lua required for benchmarking!")

local clock = os.clock

local function do_setnode(d)
	local start = clock()

	for x=0,d do
		for y=0,d do
			for z=0,d do
				minetest.set_node({x = x, y = y, z = z}, {name="default:stone"})
			end
		end
	end
	return clock() - start
end

local function do_bulkset(d)
	local start = clock()

	local poses = {}
	for x=0,d do
		for y=0,d do
			for z=0,d do
				poses[#poses + 1] = {x = x, y = y, z = z}
			end
		end
	end

	minetest.bulk_set_node(poses, {name="default:stone"})
	return clock() - start
end

local c_stone = minetest.get_content_id("default:stone")
local function do_lvm(d)
	local start = clock()

	local minp = {x=0, y=0, z=0}
	local maxp = {x=d, y=d, z=d}
	local vm = minetest.get_voxel_manip()
	local emin, emax = vm:read_from_map(minp, maxp)
	local a = VoxelArea:new{
		MinEdge = emin,
		MaxEdge = emax
	}
	local data = vm:get_data()

	for x=0,d do
		for y=0,d do
			for z=0,d do
				data[a:index(x, y, d)] = c_stone
			end
		end
	end

	vm:set_data(data)
	vm:write_to_map(data)
	vm:update_map()
	return clock() - start
end


minetest.register_chatcommand("a", {
	func = function()
		for d=10,100,10 do
			local x = do_setnode(d)
			local y = do_bulkset(d)
			local z = do_lvm(d)
			print(d .. ", " .. x .. ", " .. y .. ", " .. z)
		end
	end
})
```
</details>

**Comment by nerzhul** [comment]
@rubenwardy thanks for more precise bench, it was exactly what is said, 25% in release mode

**Comment by rubenwardy** [comment]
LVM is 95% - that's 20x rather than 1.3x

**Comment by nerzhul** [comment]
i talk a comparison between set_node and bulk_set_node. Does LVM call the node callbacks ? if not it's not really comparable

File: doc/lua_api.txt [hunk 2]
```diff
@@ -2749,6 +2749,10 @@ and `minetest.auth_reload` call the authentication handler.
     * `node`: table `{name=string, param1=number, param2=number}`
     * If param1 or param2 is omitted, it's set to `0`.
     * e.g. `minetest.set_node({x=0, y=10, z=0}, {name="default:wood"})`
+* `minetest.bulk_set_node({pos1, pos2, pos3, ...}, node)`
+    * Set node on all positions set in the first argument.
+    * e.g. `minetest.bulk_set_node({{x=0, y=1, z=1}, {x=1, y=2, z=2}}, {name="default:stone"})`
+    * For node definition or position syntax see `minetest.set_node` call
```
Comment (rubenwardy) [inline doc/lua_api.txt]: For node specification

File: doc/lua_api.txt [hunk 3]
```diff
@@ -2749,6 +2749,10 @@ and `minetest.auth_reload` call the authentication handler.
     * `node`: table `{name=string, param1=number, param2=number}`
     * If param1 or param2 is omitted, it's set to `0`.
     * e.g. `minetest.set_node({x=0, y=10, z=0}, {name="default:wood"})`
+* `minetest.bulk_set_node({pos1, pos2, pos3, ...}, node)`
+    * Set node on all positions set in the first argument.
```
Comment (rubenwardy) [inline doc/lua_api.txt]: > * Faster than set_node due to single call, but still considerably slower than Voxel Manipulators for large numbers of nodes

**Comment by sofar** [comment]
Why not expand `minetest.set_node(pos, node)` so that you can pass it multiple nodes:

`minetest.set_pos(pos1, node1, pos2, node2, pos3, node3, ...)`

File: doc/lua_api.txt (hunk 3 above)
Comment (paramat) [reply doc/lua_api.txt]: Yes please more documentation.

**Comment by paramat** [comment]
Interesting data, is set node actually faster than LVM for any value of d? It might be for very small d.

**Comment by raymoo** [comment]
The time complexity of N^3 I gave earlier was under the assumption that the user just creates a big VM covering the entire cube, rather than manually instancing one for each bloack that actually contains nodes. In that case the time complexity would be the same as bulk set node, and the extra cost would be a constant factor coming from copying the data from the blocks that aren't part of the shell.

Also an even sparser example than what I described earlier would be a wireframe box with 1-node thick edges.

@sofar 
Isn't there a limit to the number of arguments in a Lua function call? (Because of stack size)
I couldn't find any information on this by searching, but I think the stacks manipulated by the C API do have a limit.

**Comment by rubenwardy** [comment]
[graph](https://i.rubenwardy.com/JFXl3vMZZX.png)

**Comment by paramat** [comment]
I just tried small d values too.

d, number of nodes, set node time, LVM time
1, 8, 0.00028400000000062, 0.00061599999999729
2, 27, 0.00050399999999939, 0.00056999999999974
3, 64, 0.0013089999999991, 0.00055899999999909

The 2 methods are similar for a 3x3x3 volume, smaller than that set node is faster.

**Comment by nerzhul** [comment]
@rubenwardy the difference on few node is very strange as Lua API does less manipulation in every case except for node = 1. Did you use LuaJIT or regular Lua ?

**Comment by rubenwardy** [comment]
line 1 of the tests:

```lua
assert(type(jit) ~= 'table', "Plain Lua required for benchmarking!")
```

**Comment by Wuzzy2** [comment]
@rubenwardy: What is the Y axis for? I am clueless.

**Comment by rubenwardy** [comment]
Runtime in seconds, the first graph has they labeled

**Comment by nerzhul** [comment]
@rubenwardy @paramat documentation is fixes

**Comment by paramat** [comment]
Docs are good thanks.

**Comment by HybridDog** [comment]
@nerzhul, could you also optimize the lookups for remove_node?

Also, freeminer has an additional set_node parameter called "fast" (https://github.com/freeminer/freeminer/blob/master/src/environment.cpp#L1302). It can be used if you want to set a node without updating liquids, light etc. (https://github.com/freeminer/freeminer/blob/master/doc/lua_api_fm.txt#L30)
Maybe you can use this to significantly increase the speed of bulk_node_set.

**Comment by paramat** [comment]
'bulk node set' needs to update light and liquids.

**Comment by HybridDog** [comment]
It currently updates light and liquids for each node, however, the updates often happen at the same position I think, e.g. light updates need to propagate to the neighbouring nodes.

**Pull request merged**